Add FenViewer IA

rouzbeh opened the pull request:
**What does your Instant Answer do?**
This instant answer parses a FEN formatted chess position and displays it on a chessboard.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
It's a very quick way to visualise a chess position.

**What is the data source for your Instant Answer? (Provide a link if possible)**
There is no data source.

**Why did you choose this data source?**
NA

**Are there any other alternative (better) data sources?**
NA

**What are some example queries that trigger this Instant Answer?**
FEN rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Chess players

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No

**Which existing Instant Answers will this one supercede/overlap with?**
NA

**Are you having any problems? Do you need our help with anything?**
No

**Where did you hear about DuckDuckHack? (For first time contributors)**
On the DuckDuckGo main page.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
http://imgur.com/XzHecj0

## Checklist
Please place an 'X' where appropriate.

```
- [X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
- [X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
- [X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
- [X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
- [X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        - [X] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 10
        - [X] IE 11

    - Windows 7
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 8
        - [] IE 9
        - [] IE 10

    - Windows XP
        - [] IE 7
        - [] IE 8

    - Mac OSX
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] Safari

    - iOS (iPhone)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - iOS (iPad)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - Android
        - [X] Firefox
        - [] Native Browser
        - [X] Google Chrome
        - [] Opera
```

mintsoft commented on lib/DDG/Goodie/FenViewer.pm:
```diff
@@ -0,0 +1,124 @@
+zci answer_type => "fen_viewer";
+zci is_cached   => 1;
+
+# Metadata.  See https://duck.co/duckduckhack/metadata for help in filling out this section.
```
You can nuke these instructional comments

mintsoft commented on t/FenViewer.t:
```diff
@@ -0,0 +1,23 @@
+    # At a minimum, be sure to include tests for all:
+    # - primary_example_queries
+    # - secondary_example_queries
+    'FEN rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1' => test_zci('Chessboard', html => '<div class="zci--fenviewer"><table class="chess_board" cellpadding="0" cellspacing="0"><tr><td id="A8"><a href="#" class="black rook">&#9820;</a></td><td id="B8"><a href="#" class="black knight">&#9822;</a></td><td id="C8"><a href="#" class="black bishop">&#9821;</a></td><td id="D8"><a href="#" c ...
```
Given the amount of parsing in the IA can you provide some "real" test cases for non-starting positions etc?

mintsoft commented on share/goodie/fen_viewer/fen_viewer.css:
```diff
@@ -0,0 +1,35 @@
+/*
+    Document   : zeroclickinfo-goodies/share/goodie/fen_viewer/fen_viewer.css
```
This comment can also go :+1:

mintsoft commented on lib/DDG/Goodie/FenViewer.pm:
```diff
@@ -0,0 +1,124 @@
+    catch {
+        return;
+    };
+    return 'Chessboard', html => $html_out;
```
It'd be nice if the text output was an ascii representation of the board; similiar to how Sudoku does it: https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Sudoku.pm

mintsoft commented on lib/DDG/Goodie/FenViewer.pm:
```diff
@@ -0,0 +1,124 @@
+    # optional - regex guard
+    # return unless qr/^\w+/;
+    my ($query) = $_;
+    return unless $query;
```
Is it worth actually guarding this with regex? How well behaved is the parsing with invalid input?

mintsoft commented on t/FenViewer.t:
```diff
@@ -0,0 +1,23 @@
+    'fen rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq e3 0 1' => test_zci('Chessboard', html => '<div class="zci--fenviewer"><table class="chess_board" cellpadding="0" cellspacing="0"><tr><td id="A8"><a href="#" class="black rook">&#9820;</a></td><td id="B8"><a href="#" class="black knight">&#9822;</a></td><td id="C8"><a href="#" class="black bishop">&#9821;</a></td><td id="D8"><a href="# ...
+    # Try to include some examples of queries on which it might
+    # appear that your answer will trigger, but does not.
+    'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1' => undef,
```
Can you also add some queries like "fen" that should return `undef`

mintsoft commented:
@rouzbeh This looks really cool! I could have sworn we already had this or similiar implemented somewhere, however I now think I might have been getting confused with the Chess960 IA: https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Chess960.pm

Can you add a screenshot of the IA?

rouzbeh commented:
@mintsoft Thanks a lot for the review and your suggestions.

I have added ASCII output, removed useless comments, and made the parsing more robust by checking for more cases of error. There are now also more tests, including two positions from the world championship match. There was already a screenshot in the pull request. Here it is again : http://imgur.com/XzHecj0

One advantage of this IA, compared to Chess960, is that it does everything in HTML. No querying for images.

mintsoft commented on lib/DDG/Goodie/FenViewer.pm:
```diff
@@ -0,0 +1,160 @@
+# parses the current board position (the first field in the FEN format) and does not check
+# that the given position is actually legal.
+
+use DDG::Goodie;
```
Can you add `use strict;` here? It'll force github to recognise it as perl 5

mintsoft commented:
@abeyang any design feedback? http://imgur.com/XzHecj0

rouzbeh commented:
Hi

Any news on the design review ? Once this is done, I can update the Chess960 to use the same visual output instead of an image.

Cheers,
Ali

moollaza commented:
@rouzbeh sorry for the delay, we'll get around to this shortly!

abeyang commented:
Hey @rouzbeh , this looks pretty neat! Can we clean up the visuals, just a tad:

```css
.zci--answer .zci--fenviewer a {
  color: #333;
  font-size: 27px;
  line-height: 1.5;
  text-shadow: none; /* remove text-shadow altogether */
}

.zci--answer .zci--fenviewer .chess_board tr:nth-child(odd) td:nth-child(even), .zci--answer .zci--fenviewer .chess_board tr:nth-child(even) td:nth-child(odd) {
  background-color: #ddd;
  box-shadow: none; /* remove box-shadow altogether */
}

.zci--answer .zci--fenviewer .chess_board {
  border: 2px solid #333;
}
```

In the end, it should look something like this:
![screen shot 2015-03-25 at 11 03 27 am](https://cloud.githubusercontent.com/assets/794084/6831603/9aeabfdc-d2de-11e4-9394-678ee08e164c.png)

rouzbeh commented:
This is how it looks like now : 

![image](https://cloud.githubusercontent.com/assets/511375/6866388/f61e4a96-d477-11e4-8524-a1b5d055d432.png)

abeyang commented:
Hi @rouzbeh , almost there! You're still missing the `line-height: 1.5;` for `.zci--answer .zci--fenviewer a`. Other than that, it looks good!

rouzbeh commented:
@abeyang Oops, sorry about that. It should be good now.

abeyang commented:
Looks good! :+1:

moollaza commented:
LGTM :+1:

moollaza commented:
@rouzbeh congrats! It's live: https://duckduckgo.com/?q=FEN+rnbqkbnr%2Fpppppppp%2F8%2F8%2F8%2F8%2FPPPPPPPP%2FRNBQKBNR+w+KQkq+-+0+1&ia=answer

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's & Issues or even submit more Instant Answers!

rouzbeh commented:
@moollaza Hooray :)
